### ISSUE TITLE
find-cain error on |reset attempts

### ISSUE THREAD

**Issue by joshuareagan** [urbit/arvo#868]
On release-candidate I'm mounting my home desk to linux and revising hoon.hoon to add printfs.  This is what I get:

```
~zod:dojo> hot jet: /loss/hex/pen/qua/tri/two/one/k141
  axe 63, jax 11,
  bash 0adf8272a5f74f63c150b798fb3869423267665fcb4163e03eb20ad680760d31
[%vega-start-hoon /~zod/home/~2018.10.29..20.34.09..50f3/sys/hoon]
%vega-parsed
-find.cain
[%vega-kernel 0x11c.9dc4]
: /~zod/home/3/sys/hoon/hoon
```

Attempting a manual `|reset` after this point has the same result.

**Comment by ohAitch** [comment]
"cain" is the interface name for a helper function used in pretty-printing,
are you adding the printfs above the core that's defined in?

**Comment by joshuareagan** [comment]
yup, that was it.  Thanks.

Is this desired behavior?  Should I just know not to use the `~&` rune in certain places?

**Comment by ohAitch** [comment]
I think the etymology somehow involves original sin? Agreed that a single
core "--accursed-unutterable-prettyprinting-utils", which then had
"tank-vase" "port-type" etc helpers in it, would be significantly easier to
debug without context.

**Comment by joemfb** [comment]
Not sure this should be closed. It's certainly possible to imagine the vase pretty-printing being moved from the macro expansion to an explicit call in the runtime, at least for the hints. I don't know what should happen with `=>(~ >1<)` though.

**Comment by belisarius222** [comment]
Yeah. The macro expansion for `!>` also depends on having `onan` and `abel` limbs in the subject, although both of those appear to be completely useless, since they depend on `typo` and `vise`, which are just aliases for `type` and `vase` respectively.

**Comment by joshuareagan** [comment]
Vox populi....

**Comment by belisarius222** [comment]
I get the impression that to fix this, `+open:ap` and surrounding utilities would need to be reachitected to embed the `cain` gate itself into the output Nock. I assume that would be ok because of structural sharing. I'm not sure if there are downstream issues from that, though.

**Comment by ohAitch** [comment]
Keyword "appear", they're part of the type of ++type migration machinery

**Comment by cgyarvin** [comment]
This is not a bug and I should explain it.   We are definitely not
embedding any code into any output Nock.  =cain needs a very substantial
amount of code to work, obviously.

=cain and other "biblicals" -- anything in hoon.hoon that's a four-letter
Bible character -- is an exception to the usual principle that Hoon doesn't
generate references to a hoon.hoon context that may not exist -- especially
in parts of hoon.hoon above that context.  Of course you can't ~& in =add,
because ~& needs to call a whole bunch of analysis and printing code that
doesn't exist yet.

=cain just calls =sell.  We put in the extra layer of indirection to make
it as easy as possible to do interesting things with the result.  But there
is no magic here -- and we only use biblicals for non-core purposes like
this one.

**Comment by ohAitch** [comment]
I take it you are not warm on the idea of renaming  cain  to
 vase-to-tank:--accursed-unutterable-type-utils  for the sake of people who
can't find ++cain in the urbit docs?

**Comment by cgyarvin** [comment]
It's right there in the source code!

But yes, this error is confusing and could be much better explained.  I'm
not sure what the right strategy is.  However -- it will not happen to
anyone working in normal userspace with a normal subject.  If you're
already editing hoon.hoon...

**Comment by joshuareagan** [comment]
Ted's example is a lot more basic than editing hoon.hoon:

```
=>  ~  ~&  "hello"  ~
```

At the very least, I'll need to document that `~&` requires a subject that knows what a `cain` is.

**Comment by ohAitch** [comment]
Curtis does have a point that the reasonable thing to do an a -find.cain is
to well, go find ++cain: I'm not sure the urbit.org/docs or hoon.hoon
comments for it are super informative, but if not that's probably where
this information belongs.

**Comment by joshuareagan** [comment]
I guess I'm admitting to being a dummy, but when I found `cain` in hoon.hoon I had no idea what it was supposed to mean to me as a user of `~&`.

**Comment by joemfb** [comment]
The biblical names should probably be in a separate core chapter with a block level comment explaining their use. And `+typo`/`+vise` should probably have comments referencing an advanced compiler/hoon.hoon staging document (wherein someone could fully articulate the meta-vase problem). We can limit the requisite dimensions of esoteric knowledge. I believe in us.

**Comment by cgyarvin** [comment]
It’s meant to remind you that you’re being a sinner!  Unless you think Hoon is magic, how would ~& work without a subject?

I agree that it needs a real explanation...
